File: enthropic/parser.py
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


# ── helpers ───────────────────────────────────────────────────────────────────

def _strip_comment(line: str) -> str:
    idx = line.find("#")
    return line[:idx] if idx != -1 else line


def _indent(line: str) -> int:
    return len(line) - len(line.lstrip(" \t"))


def _list(s: str) -> list[str]:
    return [x.strip() for x in s.split(",") if x.strip()]


# ── data model ────────────────────────────────────────────────────────────────

@dataclass
class Transform:
    source: str
    target: str
    actions: list[str]


@dataclass
class Layer:
    name: str
    owns: list[str] = field(default_factory=list)
    can: list[str] = field(default_factory=list)
    cannot: list[str] = field(default_factory=list)
    calls: list[str] = field(default_factory=list)
    never: list[str] = field(default_factory=list)
    latency: Optional[str] = None


@dataclass
class Contract:
    subject: str
    keyword: str   # ALWAYS | NEVER | REQUIRES
    qualifier: str


@dataclass
class FlowStep:
    number: int
    subject: str
    action: str


@dataclass
class Flow:
    name: str
    steps: list[FlowStep] = field(default_factory=list)
    rollback: list[str] = field(default_factory=list)
    atomic: Optional[bool] = None
    timeout: Optional[str] = None
    retry: Optional[int] = None


@dataclass
class EnthSpec:
    source_file: str
    version: str = ""
    project: dict = field(default_factory=dict)
    vocabulary: list[str] = field(default_factory=list)
    entities: list[str] = field(default_factory=list)
    transforms: list[Transform] = field(default_factory=list)
    layers: dict[str, Layer] = field(default_factory=dict)
    contracts: list[Contract] = field(default_factory=list)
    flows: dict[str, Flow] = field(default_factory=dict)
    secrets: list[str] = field(default_factory=list)  # names only, never values


# ── parser entry point ────────────────────────────────────────────────────────

def parse(path: Path) -> EnthSpec:
    spec = EnthSpec(source_file=str(path))
    lines = path.read_text(encoding="utf-8").splitlines()
    i = 0
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        ind = _indent(clean)
        if ind > 0:
            i += 1
            continue
        # top-level keyword dispatch
        if tok.startswith("VERSION "):
            spec.version = tok[8:].strip()
            i += 1
        elif tok == "PROJECT" or tok.startswith("PROJECT "):
            if tok.startswith("PROJECT "):
                # inline identifier: PROJECT myname → use as NAME if not overridden later
                spec.project.setdefault("NAME", tok[8:].strip())
            i = _parse_project(lines, i + 1, spec)
        elif tok == "VOCABULARY":
            i = _parse_vocabulary(lines, i + 1, spec)
        elif tok.startswith("ENTITY "):
            spec.entities = _list(tok[7:])
            i += 1
        elif tok == "TRANSFORM":
            i = _parse_transform(lines, i + 1, spec)
        elif tok == "LAYERS":
            i = _parse_layers(lines, i + 1, spec)
        elif tok == "CONTRACTS":
            i = _parse_contracts(lines, i + 1, spec)
        elif tok == "SECRETS":
            i = _parse_secrets(lines, i + 1, spec)
        else:
            i += 1
    return spec


# ── block parsers ─────────────────────────────────────────────────────────────

def _parse_project(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    in_deps = False
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        ind = _indent(clean)
        if ind == 0:
            return i
        if ind <= 2:
            in_deps = (tok == "DEPS")
            if not in_deps:
                parts = tok.split(None, 1)
                if len(parts) == 2:
                    key, val = parts[0], parts[1].strip('"').strip()
                    spec.project[key] = _list(val) if key == "STACK" else val
        elif in_deps:  # indent > 2, inside DEPS
            parts = tok.split(None, 1)
            if len(parts) == 2:
                dep_key, val = parts[0], parts[1]
                if dep_key in ("SYSTEM", "RUNTIME", "DEV"):
                    existing = spec.project.setdefault("DEPS", {})
                    existing[dep_key] = _list(val)
        i += 1
    return i


def _parse_vocabulary(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        if _indent(clean) == 0:
            return i
        spec.vocabulary.append(tok.split()[0])
        i += 1
    return i


def _parse_transform(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        if _indent(clean) == 0:
            return i
        if "->" in tok and ":" in tok:
            arrow, actions = tok.split(":", 1)
            parts = arrow.split("->")
            if len(parts) == 2:
                spec.transforms.append(Transform(
                    source=parts[0].strip(),
                    target=parts[1].strip(),
                    actions=_list(actions),
                ))
        i += 1
    return i


def _parse_layers(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    current: Optional[Layer] = None
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        ind = _indent(clean)
        if ind == 0:
            return i
        if ind <= 2:   # layer name
            current = Layer(name=tok)
            spec.layers[tok] = current
        elif current:  # layer key-value
            parts = tok.split(None, 1)
            if len(parts) == 2:
                key, val = parts[0], parts[1].strip()
                if key == "OWNS":
                    current.owns = _list(val)
                elif key == "CAN":
                    current.can = _list(val)
                elif key == "CANNOT":
                    current.cannot = _list(val)
                elif key == "CALLS":
                    current.calls = _list(val)
                elif key == "NEVER":
                    current.never.append(val)
                elif key == "LATENCY":
                    current.latency = val
        i += 1
    return i


def _parse_contracts(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    current_flow: Optional[Flow] = None
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        ind = _indent(clean)
        if ind == 0:
            return i
        if ind <= 2:
            if tok.startswith("FLOW "):
                name = tok[5:].strip()
                current_flow = Flow(name=name)
                spec.flows[name] = current_flow
            else:
                current_flow = None
                parts = tok.split(None, 2)
                if len(parts) == 3:
                    subj, kw, qual = parts
                    if kw in ("ALWAYS", "NEVER", "REQUIRES"):
                        spec.contracts.append(Contract(subject=subj, keyword=kw, qualifier=qual))
        elif current_flow:
            tokens = tok.split(None, 1)
            if not tokens:
                i += 1
                continue
            first = tokens[0]
            rest = tokens[1].strip() if len(tokens) > 1 else ""
            # flow step: "N. subject.action"
            if first.endswith(".") and first[:-1].isdigit():
                num = int(first[:-1])
                if "." in rest:
                    subj, act = rest.split(".", 1)
                    current_flow.steps.append(FlowStep(number=num, subject=subj.strip(), action=act.strip()))
                else:
                    current_flow.steps.append(FlowStep(number=num, subject="", action=rest))
            # flow meta
            elif first == "ROLLBACK":
                current_flow.rollback = _list(rest)
            elif first == "ATOMIC":
                current_flow.atomic = rest.lower() == "true"
            elif first == "TIMEOUT":
                current_flow.timeout = rest
            elif first == "RETRY":
                try:
                    current_flow.retry = int(rest)
                except ValueError:
                    pass
        i += 1
    return i


def _parse_secrets(lines: list[str], start: int, spec: EnthSpec) -> int:
    i = start
    while i < len(lines):
        clean = _strip_comment(lines[i])
        tok = clean.strip()
        if not tok:
            i += 1
            continue
        if _indent(clean) == 0:
            return i
        # first token is the key name — value is never stored here
        spec.secrets.append(tok.split()[0])
        i += 1
    return i

File: src/parser.ts
import { readFileSync } from 'fs';
import { resolve } from 'path';

export interface Transform {
  source: string;
  target: string;
  actions: string[];
}

export interface Layer {
  name: string;
  owns: string[];
  can: string[];
  cannot: string[];
  calls: string[];
  never: string[];
  boundary?: string;
  exposes: string[];
  latency?: string;
}

export interface Contract {
  subject: string;
  keyword: string;
  qualifier: string;
}

export interface FlowStep {
  number: number;
  layer?: string;
  subject: string;
  action: string;
}

export interface Flow {
  name: string;
  steps: FlowStep[];
  rollback: string[];
  atomic?: boolean;
  timeout?: string;
  retry?: number;
}

export type ProjectValue =
  | { kind: 'str'; value: string }
  | { kind: 'list'; value: string[] }
  | { kind: 'deps'; value: Record<string, string[]> };

export interface OwnershipEntry {
  kind: 'entity' | 'flow';
  name: string;
  owners: string[];          // e.g. ["alice@company"]
  requiredReview: string[];  // e.g. ["alice", "bob"] if "required-review" present
}

export interface ObservabilityEntry {
  flow: string;
  level?: string;          // critical | warn | info
  mustLog: string[];
  mustNotLog: string[];
  metrics: string[];
}

export interface TestingEntry {
  flow: string;
  coverage?: number;       // percentage
  requiredTests: string[];
}

export interface QuotaEntry {
  resource: string;   // e.g. "stripe.charges"
  limit: string;      // e.g. "1000/day"
}

export interface ChangelogEntry {
  keyword: 'BREAKING' | 'ADDED' | 'DEPRECATED' | 'CHANGED';
  description: string;
}

export interface ChangelogVersion {
  version: string;
  entries: ChangelogEntry[];
}

export interface ClassifyEntry {
  field: string;
  class: 'credential' | 'pii' | 'sensitive' | 'internal';
}

export interface PerformanceEntry {
  entity: string;
  cardinality?: string;
  hotspot?: string;
  requires: string[];
  baseline?: {
    p50?: string;
    p99?: string;
    maxMemory?: string;
  };
  constraints?: {
    p99?: string;
    errorRate?: string;
  };
}

export interface SpeqSpec {
  sourceFile: string;
  version: string;
  project: Map<string, ProjectValue>;
  vocabulary: string[];
  entities: string[];
  transforms: Transform[];
  layers: Map<string, Layer>;
  layersOrder: string[];
  contracts: Contract[];
  flows: Map<string, Flow>;
  flowsOrder: string[];
  secrets: string[];
  ownership: OwnershipEntry[];
  observability: Map<string, ObservabilityEntry>;
  testing: Map<string, TestingEntry>;
  quotas: QuotaEntry[];
  performance: PerformanceEntry[];
  changelog: ChangelogVersion[];
  classify: ClassifyEntry[];
  secretScopes: Map<string, string>;
}

function stripComment(line: string): string {
  const idx = line.indexOf('#');
  return idx === -1 ? line : line.slice(0, idx);
}

function indentLen(line: string): number {
  return line.length - line.trimStart().length;
}

export function splitList(s: string): string[] {
  return s.split(',').map(x => x.trim()).filter(x => x.length > 0);
}

const BLOCK_INDENT = 2;

function defaultSpec(sourceFile: string): SpeqSpec {
  return {
    sourceFile: resolve(sourceFile),
    version: '',
    project: new Map(),
    vocabulary: [],
    entities: [],
    transforms: [],
    layers: new Map(),
    layersOrder: [],
    contracts: [],
    flows: new Map(),
    flowsOrder: [],
    secrets: [],
    ownership: [],
    observability: new Map(),
    testing: new Map(),
    quotas: [],
    performance: [],
    changelog: [],
    classify: [],
    secretScopes: new Map(),
  };
}

export function parse(path: string): SpeqSpec {
  const content = readFileSync(path, 'utf-8');
  const lines = content.split('\n');
  const spec = defaultSpec(path);

  let i = 0;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind > 0) { i++; continue; }

    if (tok.startsWith('VERSION ')) {
      spec.version = tok.slice('VERSION '.length).trim();
      i++;
    } else if (tok === 'PROJECT' || tok.startsWith('PROJECT ')) {
      if (tok.startsWith('PROJECT ')) {
        const name = tok.slice('PROJECT '.length).trim();
        if (!spec.project.has('NAME')) {
          spec.project.set('NAME', { kind: 'str', value: name });
        }
      }
      i = parseProject(lines, i + 1, spec);
    } else if (tok === 'CONTEXT') {
      i = parseContext(lines, i + 1, spec);
    } else if (tok === 'VOCABULARY') {
      i = parseVocabulary(lines, i + 1, spec);
    } else if (tok.startsWith('ENTITY ')) {      // inline: ENTITY foo, bar, baz
      spec.entities = splitList(tok.slice('ENTITY '.length));
      i++;
    } else if (tok === 'ENTITY') {
      // multiline: indented lines follow
      i = parseEntityMultiline(lines, i + 1, spec);
    } else if (tok === 'TRANSFORM') {
      i = parseTransform(lines, i + 1, spec);
    } else if (tok === 'LAYERS') {
      i = parseLayers(lines, i + 1, spec);
    } else if (tok === 'CONTRACTS') {
      i = parseContracts(lines, i + 1, spec);
    } else if (tok === 'SECRETS') {
      i = parseSecrets(lines, i + 1, spec);
    } else if (tok === 'OWNERSHIP') {
      i = parseOwnership(lines, i + 1, spec);
    } else if (tok === 'OBSERVABILITY') {
      i = parseObservability(lines, i + 1, spec);
    } else if (tok === 'TESTING') {
      i = parseTesting(lines, i + 1, spec);
    } else if (tok === 'QUOTAS') {
      i = parseQuotas(lines, i + 1, spec);
    } else if (tok === 'PERFORMANCE') {
      i = parsePerformance(lines, i + 1, spec);
    } else if (tok === 'CLASSIFY') {
      i = parseClassify(lines, i + 1, spec);
    } else if (tok === 'CHANGELOG') {
      i = parseChangelog(lines, i + 1, spec);
    } else {
      i++;
    }
  }

  return spec;
}

function parseEntityMultiline(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    // Each indented non-empty non-comment line is one entity name
    const name = tok.split(/\s+/)[0];
    if (name) spec.entities.push(name);
    i++;
  }
  return i;
}

function parseProject(lines: string[], start: number, spec: SpeqSpec): number {
  let i = start;
  let inDeps = false;
  const depsMap: Record<string, string[]> = {};

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind === 0) {
      if (Object.keys(depsMap).length > 0) {
        spec.project.set('DEPS', { kind: 'deps', value: depsMap });
      }
      return i;
    }
    if (ind <= BLOCK_INDENT) {
      inDeps = tok === 'DEPS';
      if (!inDeps) {
        const spaceIdx = tok.search(/\s/);
        if (spaceIdx !== -1) {
          const key = tok.slice(0, spaceIdx).trim();
          const val = tok.slice(spaceIdx + 1).trim().replace(/^"|"$/g, '').trim();
          if (key === 'STACK') {
            spec.project.set(key, { kind: 'list', value: splitList(val) });
          } else {
            spec.project.set(key, { kind: 'str', value: val });
          }
        }
      }
    } else if (inDeps) {
      const spaceIdx = tok.search(/\s/);
      if (spaceIdx !== -1) {
        const depKey = tok.slice(0, spaceIdx).trim();
        const val = tok.slice(spaceIdx + 1).trim();
        if (depKey === 'SYSTEM' || depKey === 'RUNTIME' || depKey === 'DEV') {
          depsMap[depKey] = splitList(val);
        }
      }
    }
    i++;
  }
  if (Object.keys(depsMap).length > 0) {
    spec.project.set('DEPS', { kind: 'deps', value: depsMap });
  }
  return i;
}

function parseVocabulary(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    const first = tok.split(/\s+/)[0];
    if (first) spec.vocabulary.push(first);
    i++;
  }
  return i;
}

function parseTransform(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    if (tok.includes('->') && tok.includes(':')) {
      const colonIdx = tok.indexOf(':');
      const arrow = tok.slice(0, colonIdx);
      const actionsStr = tok.slice(colonIdx + 1);
      const parts = arrow.split('->');
      if (parts.length === 2) {
        spec.transforms.push({
          source: parts[0].trim(),
          target: parts[1].trim(),
          actions: splitList(actionsStr),
        });
      }
    }
    i++;
  }
  return i;
}

function parseLayers(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  let current: string | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind <= stopIndent) return i;

    if (ind <= stopIndent + BLOCK_INDENT) {
      const name = tok;
      current = name;
      spec.layersOrder.push(name);
      spec.layers.set(name, { name, owns: [], can: [], cannot: [], calls: [], never: [], exposes: [] });
    } else if (current !== null) {
      const spaceIdx = tok.search(/\s/);
      if (spaceIdx !== -1) {
        const key = tok.slice(0, spaceIdx).trim();
        const val = tok.slice(spaceIdx + 1).trim();
        const layer = spec.layers.get(current)!;
        switch (key) {
          case 'OWNS': layer.owns = splitList(val); break;
          case 'CAN': layer.can = splitList(val); break;
          case 'CANNOT': layer.cannot = splitList(val); break;
          case 'CALLS': layer.calls = splitList(val); break;
          case 'NEVER': layer.never.push(val); break;
          case 'BOUNDARY': layer.boundary = val; break;
          case 'EXPOSES': layer.exposes = splitList(val); break;
          case 'LATENCY': layer.latency = val; break;
        }
      }
    }
    i++;
  }
  return i;
}

function parseContext(lines: string[], start: number, spec: SpeqSpec): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind === 0) return i;

    if (ind <= BLOCK_INDENT) {
      if (tok === 'VOCABULARY') {
        i = parseVocabulary(lines, i + 1, spec, BLOCK_INDENT);
      } else if (tok.startsWith('ENTITY ')) {
        spec.entities = splitList(tok.slice('ENTITY '.length));
        i++;
      } else if (tok === 'ENTITY') {
        i = parseEntityMultiline(lines, i + 1, spec, BLOCK_INDENT);
      } else if (tok === 'TRANSFORM') {
        i = parseTransform(lines, i + 1, spec, BLOCK_INDENT);
      } else if (tok === 'LAYERS') {
        i = parseLayers(lines, i + 1, spec, BLOCK_INDENT);
      } else if (tok === 'OWNERSHIP') {
        i = parseOwnership(lines, i + 1, spec, BLOCK_INDENT);
      } else if (tok === 'PERFORMANCE') {
        i = parsePerformance(lines, i + 1, spec, BLOCK_INDENT);
      } else {
        i++;
      }
    } else {
      i++;
    }
  }
  return i;
}

function parseContracts(lines: string[], start: number, spec: SpeqSpec): number {
  let i = start;
  let currentFlow: string | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind === 0) return i;

    if (ind <= BLOCK_INDENT) {
      if (tok === 'QUOTAS') {
        currentFlow = null;
        i = parseQuotas(lines, i + 1, spec, BLOCK_INDENT);
        continue;
      } else if (tok === 'OBSERVABILITY') {
        currentFlow = null;
        i = parseObservability(lines, i + 1, spec, BLOCK_INDENT);
        continue;
      } else if (tok === 'TESTING') {
        currentFlow = null;
        i = parseTesting(lines, i + 1, spec, BLOCK_INDENT);
        continue;
      } else if (tok === 'CHANGELOG') {
        currentFlow = null;
        i = parseChangelog(lines, i + 1, spec, BLOCK_INDENT);
        continue;
      } else if (tok.startsWith('FLOW ')) {
        const name = tok.slice('FLOW '.length).trim();
        currentFlow = name;
        spec.flowsOrder.push(name);
        spec.flows.set(name, { name, steps: [], rollback: [] });
      } else {
        currentFlow = null;
        const parts = tok.split(/\s+/);
        if (parts.length >= 3) {
          const [subj, kw, ...rest] = parts;
          if (kw === 'ALWAYS' || kw === 'NEVER' || kw === 'REQUIRES') {
            spec.contracts.push({ subject: subj, keyword: kw, qualifier: rest.join(' ') });
          }
        }
      }
    } else if (currentFlow !== null) {
      const spaceIdx = tok.search(/\s/);
      const first = spaceIdx === -1 ? tok : tok.slice(0, spaceIdx);
      const rest = spaceIdx === -1 ? '' : tok.slice(spaceIdx + 1).trim();

      if (first.endsWith('.') && /^\d+$/.test(first.slice(0, -1))) {
        const num = parseInt(first.slice(0, -1), 10);
        const flow = spec.flows.get(currentFlow)!;
        let stepText = rest;
        let stepLayer: string | undefined;
        const layerTagMatch = stepText.match(/^\[([A-Z][A-Z0-9_]*)\]\s*/);
        if (layerTagMatch) {
          stepLayer = layerTagMatch[1];
          stepText = stepText.slice(layerTagMatch[0].length);
        }
        const dotIdx = stepText.indexOf('.');
        if (dotIdx !== -1) {
          flow.steps.push({ number: num, layer: stepLayer, subject: stepText.slice(0, dotIdx).trim(), action: stepText.slice(dotIdx + 1).trim() });
        } else {
          flow.steps.push({ number: num, layer: stepLayer, subject: '', action: stepText });
        }
      } else {
        const flow = spec.flows.get(currentFlow);
        if (flow) {
          switch (first) {
            case 'ROLLBACK': flow.rollback = splitList(rest); break;
            case 'ATOMIC': flow.atomic = rest.toLowerCase() === 'true'; break;
            case 'TIMEOUT': flow.timeout = rest; break;
            case 'RETRY': {
              const n = parseInt(rest, 10);
              if (!isNaN(n)) flow.retry = n;
              break;
            }
          }
        }
      }
    }
    i++;
  }
  return i;
}

function parseSecrets(lines: string[], start: number, spec: SpeqSpec): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) === 0) return i;
    const arrowIdx = tok.indexOf('->');
    const keyPart = (arrowIdx !== -1 ? tok.slice(0, arrowIdx) : tok).trim();
    const key = keyPart.split(/\s+/)[0];
    if (key) {
      spec.secrets.push(key);
      if (arrowIdx !== -1) {
        const layer = tok.slice(arrowIdx + 2).trim().split(/\s+/)[0];
        if (layer) spec.secretScopes.set(key, layer);
      }
    }
    i++;
  }
  return i;
}

function parseOwnership(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    // format: (entity|flow) <name> -> <owners> [(required-review)]
    const m = tok.match(/^(entity|flow)\s+(\S+)\s+->\s+(.+)$/);
    if (m) {
      const kind = m[1] as 'entity' | 'flow';
      const name = m[2];
      const rest = m[3].trim();
      const hasRequired = rest.includes('(required-review)');
      const ownersStr = rest.replace(/\(required-review\)/, '').trim().replace(/,\s*$/, '');
      const owners = splitList(ownersStr);
      spec.ownership.push({
        kind,
        name,
        owners,
        requiredReview: hasRequired ? owners.slice() : [],
      });
    }
    i++;
  }
  return i;
}

function parseObservability(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  let current: string | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind <= stopIndent) return i;

    if (ind <= stopIndent + BLOCK_INDENT) {
      if (tok.startsWith('flow ')) {
        const name = tok.slice('flow '.length).trim();
        current = name;
        spec.observability.set(name, { flow: name, mustLog: [], mustNotLog: [], metrics: [] });
      }
    } else if (current !== null) {
      const colonIdx = tok.indexOf(':');
      if (colonIdx !== -1) {
        const key = tok.slice(0, colonIdx).trim().toLowerCase();
        const val = tok.slice(colonIdx + 1).trim();
        const entry = spec.observability.get(current)!;
        switch (key) {
          case 'level': entry.level = val; break;
          case 'must-log': entry.mustLog = splitList(val); break;
          case 'must-not-log': entry.mustNotLog = splitList(val); break;
          case 'metrics': entry.metrics = splitList(val); break;
        }
      }
    }
    i++;
  }
  return i;
}

function parseTesting(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  let current: string | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind <= stopIndent) return i;

    if (ind <= stopIndent + BLOCK_INDENT) {
      if (tok.startsWith('flow ')) {
        const name = tok.slice('flow '.length).trim();
        current = name;
        spec.testing.set(name, { flow: name, requiredTests: [] });
      }
    } else if (current !== null) {
      const colonIdx = tok.indexOf(':');
      if (colonIdx !== -1) {
        const key = tok.slice(0, colonIdx).trim().toLowerCase();
        const val = tok.slice(colonIdx + 1).trim();
        const entry = spec.testing.get(current)!;
        switch (key) {
          case 'coverage': {
            const n = parseInt(val, 10);
            if (!isNaN(n)) entry.coverage = n;
            break;
          }
          case 'required-tests': entry.requiredTests = splitList(val); break;
        }
      }
    }
    i++;
  }
  return i;
}

function parseQuotas(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    const colonIdx = tok.indexOf(':');
    if (colonIdx !== -1) {
      const resource = tok.slice(0, colonIdx).trim();
      const limit = tok.slice(colonIdx + 1).trim();
      if (resource) spec.quotas.push({ resource, limit });
    }
    i++;
  }
  return i;
}

function parsePerformance(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  let current: PerformanceEntry | null = null;
  let subBlock: 'baseline' | 'constraints' | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind <= stopIndent) return i;

    if (ind <= stopIndent + BLOCK_INDENT) {
      // "entity <name>"
      if (tok.startsWith('entity ')) {
        const name = tok.slice('entity '.length).trim();
        current = { entity: name, requires: [] };
        spec.performance.push(current);
        subBlock = null;
      }
    } else if (ind <= stopIndent + BLOCK_INDENT * 2 && current !== null) {
      // flat key-value OR sub-block keyword
      const colonIdx = tok.indexOf(':');
      if (colonIdx !== -1) {
        const key = tok.slice(0, colonIdx).trim().toLowerCase();
        const val = tok.slice(colonIdx + 1).trim();
        subBlock = null;
        switch (key) {
          case 'cardinality': current.cardinality = val; break;
          case 'hotspot': current.hotspot = val; break;
          case 'requires': current.requires = splitList(val); break;
        }
      } else {
        // sub-block keyword (baseline / constraints)
        const kw = tok.toLowerCase();
        if (kw === 'baseline' || kw === 'constraints') {
          subBlock = kw as 'baseline' | 'constraints';
          if (subBlock === 'baseline' && !current.baseline) current.baseline = {};
          if (subBlock === 'constraints' && !current.constraints) current.constraints = {};
        }
      }
    } else if (ind <= stopIndent + BLOCK_INDENT * 3 && current !== null && subBlock !== null) {
      // key-value inside sub-block
      const colonIdx = tok.indexOf(':');
      if (colonIdx !== -1) {
        const key = tok.slice(0, colonIdx).trim().toLowerCase();
        const val = tok.slice(colonIdx + 1).trim();
        if (subBlock === 'baseline') {
          if (!current.baseline) current.baseline = {};
          switch (key) {
            case 'p50': current.baseline.p50 = val; break;
            case 'p99': current.baseline.p99 = val; break;
            case 'max-memory': current.baseline.maxMemory = val; break;
          }
        } else if (subBlock === 'constraints') {
          if (!current.constraints) current.constraints = {};
          switch (key) {
            case 'p99': current.constraints.p99 = val; break;
            case 'error-rate': current.constraints.errorRate = val; break;
          }
        }
      }
    }
    i++;
  }
  return i;
}

const VALID_CLASSIFY_CLASSES = new Set(['credential', 'pii', 'sensitive', 'internal']);

function parseClassify(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    if (indentLen(clean) <= stopIndent) return i;
    const parts = tok.split(/\s+/);
    if (parts.length >= 2) {
      const cls = parts[1].toLowerCase();
      if (VALID_CLASSIFY_CLASSES.has(cls)) {
        spec.classify.push({ field: parts[0], class: cls as ClassifyEntry['class'] });
      }
    }
    i++;
  }
  return i;
}

const VALID_CHANGELOG_KEYWORDS = new Set(['BREAKING', 'ADDED', 'DEPRECATED', 'CHANGED']);

function parseChangelog(lines: string[], start: number, spec: SpeqSpec, stopIndent = 0): number {
  let i = start;
  let currentVersion: ChangelogVersion | null = null;

  while (i < lines.length) {
    const clean = stripComment(lines[i]);
    const tok = clean.trim();
    if (!tok) { i++; continue; }
    const ind = indentLen(clean);
    if (ind <= stopIndent) return i;

    if (ind <= stopIndent + BLOCK_INDENT) {
      // version string e.g. "v0.2.0"
      currentVersion = { version: tok, entries: [] };
      spec.changelog.push(currentVersion);
    } else if (currentVersion !== null) {
      // "KEYWORD description"
      const spaceIdx = tok.search(/\s/);
      if (spaceIdx !== -1) {
        const kw = tok.slice(0, spaceIdx).trim();
        const desc = tok.slice(spaceIdx + 1).trim();
        if (VALID_CHANGELOG_KEYWORDS.has(kw)) {
          currentVersion.entries.push({
            keyword: kw as ChangelogEntry['keyword'],
            description: desc,
          });
        }
      }
    }
    i++;
  }
  return i;
}
